odl/solvers/iterative/statistical.py

```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
+        Right-hand sides of the equation defining the inverse problem.
+    niter : int, optional
+        Number of iterations.
+    noise : {'poisson'}, optional
+        Noise model determining the variant of MLEM.
+        For ``'poisson'``, the initial value of ``x`` should be
+        non-negative.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+
+    Other Parameters
+    ----------------
+    sensitivities : float or ``op.domain`` `element-like`, optional
+        Usable with ``noise='poisson'``. The algorithm contains a ``A^T 1``
+        term, if this parameter is given, it is replaced by it.
+        Default: ``op[i].adjoint(op[i].range.one())``
+
+    Notes
+    -----
+    Given a forward models :math:`A_i`, data :math:`g_i`, :math:`i = 1, ..., M`
+    and noise model :math:`X`, the algorithm attempts find an :math:`x` that
+    maximizes:
+
+    .. math::
+
+        \prod_{i=1}^M P(g_i | g_i \\text{ is } X(A_i(x)) \\text{ distributed})
+
+    where the expectation of :math:`X(A_i(x))` satisfies
+
+    .. math::
+
+        \\mathbb{E}(X(A_i(x))) = A_i(x)
+
+    with 'poisson' noise the algorithm is given by partial updates:
+
+    .. math::
+
+       x_{n + m/M} =
+       \\frac{x_{n + (m - 1)/M}}{A_i^* 1} A_i^* (g_i / A_i(x_{n + (m - 1)/M}))
+
+    for :math:`m = 1, ..., M` and :math:`x_{n+1} = x_{n + M/M}`
+
+    See Also
+    --------
+    mlem : Ordinary MLEM algorithm without subsets.
+    loglikelihood : Function for calculating the logarithm of the likelihood
+    """
+    noise, noise_in = str(noise).lower(), noise
+    if noise not in AVAILABLE_MLEM_NOISE:
+        raise NotImplemented("noise '{}' not understood"
+                             ''.format(noise_in))
+
+    n_ops = len(op)
+    if len(data) != n_ops:
+        raise ValueError('number of data does not match number of operators')
+    if not all(x in opi.domain for opi in op):
+        raise ValueError('`x` not an element in the domains of all operators')
+
+    # Convert data to range elements
+    data = [op[i].range.element(data[i]) for i in range(len(op))]
+
+    if noise == 'poisson':
+        # Parameter used to enforce positivity.
+        # TODO: let users give this.
+        eps = 1e-8
+
+        if np.any(np.less(x, 0)):
+            raise ValueError('`x` must be non-negative')
+
+        # Extract the sensitivites parameter
+        sensitivities = kwargs.pop('sensitivities', None)
+        if sensitivities is None:
+            sensitivities = [np.maximum(opi.adjoint(opi.range.one()), eps)
+                             for opi in op]
+        else:
+            # Make sure the sensitivities is a list of the correct size.
+            try:
+                list(sensitivities)
+            except TypeError:
+                sensitivities = [sensitivities] * n_ops
+
+        tmp_dom = op[0].domain.element()
+        tmp_ran = [opi.range.element() for opi in op]
```

Comment:
This has the same memory use as the regular MLEM algorithm. I'll wait with that until later.

I tried this for the arithmetic operators and it turned out to simply be a lot of work and no one has used it so far.